### ISSUE TITLE
Note 'replies-to' should be the object ID of another note

### ISSUE THREAD

**Issue by ngriffiths21** [ngriffiths21/emr#11]

**Comment by ngriffiths21** [comment]
Replying to a note should include the object ID of the note, and also an optional field for the section of the previous note that is being addressed.

**Comment by zbo14** [comment]
> the section of the previous note that is being addressed.

Section of the previous note's data?

**Comment by ngriffiths21** [comment]
Yeah, to allow clarification of which part is being replied to. In general this will come in some kind of "plan" section at the end of the data.

**Comment by zbo14** [comment]
Ok so `note.data` is currently a string. Should the optional field indicate a range in the data it's referring to? (e.g. [0, 10]). What if it wants to refer to multiple sections of the data, disparate yet related, or is that not a use case?

**Comment by ngriffiths21** [comment]
Yeah, that wouldn't be bad to have. While we're at it we may as well allow
for replying to multiple notes. So it would really be an array, where each
entry references a note as well as a range within that note, but the range
is optional and if left out the whole note is being referenced.

Make sense?

**Comment by zbo14** [comment]
What about making data an object? Then instead of having to deal with ranges notes can reply to certain keys in previous notes?

**Comment by ngriffiths21** [comment]
This is looking ahead to a UI thing. The entire data section will come from
a single text box. I figured it was more simple for the db to match the way
it's implemented.

**Comment by zbo14** [comment]
Okay first, how something is stored in the db is not necessarily how it appears in the ui (nor should it be). That being said, we can write markup (or code that generates markup) from a data object to make it *appear* as a single text box to the user (with (sub)headings for the different (nested) keys)

**Comment by ngriffiths21** [comment]
I agree that it may not be best to have a single data string, but I'm not sure of the best way to represent the data.

Here are some thoughts I have:

- The main parts of the note we care about are (1) the entire contents of `data`, for display purposes, and (2) any section that is being flagged or is being replied to.
- It's definitely preferable to access flagged/replied sections directly rather than using a range.
- The contents, location, and length of flagged/replied sections should be as flexible as possible.
- Flagged/replied sections can be added or taken away at any time after note creation. There can be a bunch of them, or none.
- More broadly, the app **should not enforce any particular structure** for the contents of the note (`data`). This is super important to me, it's one of the main reasons I wanted to do this project in the first place.

So given all this, I had basically given up and left it as a single field.

Would it make sense to use an array of strings, such that flagged sections could be indexed? Any other ideas that would satisfy all the requirements?

**Comment by zbo14** [comment]
I don't totally understand what you mean by flagged.. Does a note flag sections of another note?

> the app should not enforce any particular structure for the contents of the note (`data`)

In that case, I think the `data` should be an object. Joi will *only* check that it's an object. The physician writing the note can choose what keys/types of values are in the `data` (the ui could provide a slate of options with ability to customize).

If you're replying to a section in a previous `note`'s data, I think `note.data.diagnosis` is clearer than `note.data[1]` or `note.data.slice(10, 20)`.

**Comment by ngriffiths21** [comment]
Check out #9 for more description of flags. Basically, yes, it flags a section of a note for someone else's review.

> Joi will only check that it's an object. The physician writing the note can choose what keys/types of values are in the data

To me, this doesn't satisfy the constraint. It enforces a particular structure (albeit a very flexible one) where the note is made up of keys and values.

The top priority here is to protect the experience of writing a note -- so that it is like writing in a journal, and not at all like filling out paperwork. The idea of physicians choosing keys for particular sections is exactly the kind of thing I'd like to avoid if at all possible. It's ideal if one physician can use headings like "SUBJECTIVE," and another can use "S:", and another can call it "HPI," and another can not use any headings at all and just write in paragraph form (I have seen each of these many times). I want it to be valid to write "Patient is doing well today," and equally valid to write a very long note with a detailed history, exam, medication list, lab results, assessment, and plan. I don't think the object form can handle this amount of variability, at least not in the way I think you are proposing. Again, this is something I feel very strongly about and want to get right.

> I think note.data.diagnosis is clearer

This is totally a trade-off. But in my opinion there are very few objects that the system itself needs to be aware of. They include flags, assessment (the physician's diagnoses and other conclusions), and the results of workup (labs, imaging etc). The vast majority of the note can (and therefore should) be handled as raw text.

**Comment by zbo14** [comment]
> It's ideal if one physician can use headings like "SUBJECTIVE," and another can use "S:", and another can call it "HPI," and another can not use any headings at all and just write in paragraph form (I have seen each of these many times). I want it to be valid to write "Patient is doing well today," and equally valid to write a very long note with a detailed history, exam, medication list, lab results, assessment, and plan.

Just thought I'd write up some examples. If physicians want to create their own headings those can be the keys of the object, and subheadings could be nested keys. If a physician wants to write a single block of text (or maybe just a sentence) `note.data` could be a string. If they want to break the text into sections, it could be an array of strings. We can allow multiple data types, but then we have to handle different types of accessing/indexing when a physician is trying to reply to the note (e.g. `note.data[2]` if it's an array or `note.data.exam` if it's an object). I think this is alright? just something to keep in mind.

Note: the following examples are db documents. This is not what the physician will be filling out in the ui! I shouldn't have said they choose the keys because that's a db concern not a ui concern. But if they want to use headings it makes sense that those would map to keys in a `note.data` object, no?

Curious to hear your thoughts/see your edits.

```js
// DB document examples

// Note 1
{
  data: {
    SUBJECTIVE: 'Patient is doing well today'
  }
}

// Note 2
{
  data: {
    S: 'Patient is doing well today'
  }
}

// Note 3
{
  data: 'Patient is doing well today'
}

// Note 4
{
  data: [ 'Paragraph 1', 'Paragraph 2', ... ]
}

// Note 5
{
  data: {
    history: { ... },
    exam: { ... },
    medication: [ ... ],
    ...
  }
}
```

**Comment by ngriffiths21** [comment]
I think you're saying we could allow any of these to be valid and to coexist in the db, and that we could simply translate the raw text block from the UI into objects of these various formats. Is that right?

First, is there any way to for all these to be valid? In one case data is a string, in another it is an object, etc. Does Joi consider a string or array to be an object? If so that answers the question of validation. Flexibility would be great here.

In general though, I'm not expecting it to be particularly useful to do this. I mentioned that flagged sections, assessment, and lab results are the only truly important sections. The rest is just other stuff. Why is it useful to separate the subjective from the objective or label them at all? Also, they aren't ordered, right? Order needs to be preserved, and these sections may not always be ordered in the same way.

(By the way, [here](https://www.nurseone.ca/~/media/nurseone/page-content/pdf-en/soap_documentation_e.pdf?la=en) is a nice description of SOAP note format which I talk about a lot).

**Comment by zbo14** [comment]
> First, is there any way to for all these to be valid?

Yes, we can use `Joi.any()` or [this](https://github.com/hapijs/joi/blob/master/API.md#alternativestryschemas).

> I mentioned that flagged sections, assessment, and lab results are the only truly important sections.

I'm looking at the note schema right now and I see there is an assessment field. Would the lab results live under the note data or would they have their own field?

**Comment by ngriffiths21** [comment]
They would be under note data. So does assessment — the assessment field
currently just points to the actual assessment.

Many notes will not need a lab results section, I guess that’s why I didn’t
include it at first.

**Comment by zbo14** [comment]
Ahh I see..

So you want the SOAP information to be in `note.data`. Ideally `note.data` is an object where each heading is a key (e.g. 'subjective') and the values are strings (e.g. 'Patient is doing well today'). The catch is you don't want the physician to fill out (or feel like they're filling out) a SOAP form.. They should be able to jot down whatever they want. Then the task is mapping what the physician writes to SOAP format (if possible). Or do you not want to follow SOAP?

What about in the UI we have different sections for S, O, A, and P? (possibly with different headings)

**Comment by ngriffiths21** [comment]
> Ideally note.data is an object where each heading is a key

What makes this ideal?

> Then the task is mapping what the physician writes to SOAP format

I feel like it would be a lot of work (see your example notes, all of which can be valid in the same implementation), and I'm not convinced it's necessary. See what I wrote above:

> I mentioned that flagged sections, assessment, and lab results are the only truly important sections. The rest is just other stuff. Why is it useful to separate the subjective from the objective or label them at all? Also, they aren't ordered, right? Order needs to be preserved, and these sections may not always be ordered in the same way.

and even if it were useful to label sections:

> The contents, location, and length of flagged/replied sections [and any other section] should be as flexible as possible.

So in other words, I don't think it's feasible to label them without making the user do it. I'm not totally against labels -- flags are labels, and they are totally necessary -- but right now I have no interest in labeling S or O. They should just be a big block of text at the beginning.

**Comment by ngriffiths21** [comment]
All this is why I liked the idea of an array:

```
{
  data: [
    'S: a whole bunch of stuff that includes subjective, O: objective, etc',
    'a lab result',
    'some more random objective stuff... and at this point there's an A: for assessment... more stuff...',
    'here is the part I actually want to flag as this note's assessment',
    'P: then I add some plan items that I don't want to flag yet',
    'here's one I do want to flag',
    'here's another flag'
  ]
}
```

See how there are important parts, unimportant parts, a specific order, and no straightforward way of labeling everything?
